### PR TITLE
Implemented DCT decompression

### PR DESCRIPTION
**Pull request by wolfhechel**
Added a functional DCT decompressor instead of using dct.dll, have not been able to test Unpack_Bloc properly.

There's also a commit which implements some fixes needed for compiling on Linux.

### CODE REVIEW

**Comment by wolfhechel** [comment]
@Alys-Elica this isn't quite ready for merge yet, but it works and I wanted to keep you in the loop and get some early feedback.

I'm much more of a C developer than C++ so if I'm doing anything weird could you perhaps leave a comment there?

Also, I haven't finished writing up how the DCT works, I'll add some documentation and more comments before merging.

One thing you'll notice if you compare this implementation with the previous one is that the color saturations are way higher. My suspicion is that the floating to long conversion is different from that of msvcrt, the change happens in the YCbCr -> RGBA conversion.

Also I'm thinking that perhaps its a good idea to retain the dll based implementation that was already in place and male it conditional, what are your thoughts on that?

File: Tools/LibFVR/internal/dct.cpp [hunk 1]
```diff
@@ -1,107 +1,598 @@
...
+        /* When packing RGB565 into our data we're left with some 
+           overflowing bits. These are contributed back into the next pixel.
+
+           The algorithm unpacks a total of width * 8 per iteration, so
+           we'll keep an extra width size for the overflowing bits which
+           will pass on to the next iteration. */
+        
+        int *r = (int *)malloc(sizeof(int[width * 9])),
+            *g = (int *)malloc(sizeof(int[width * 9])),
+            *b = (int *)malloc(sizeof(int[width * 9]));
```
Comment (Alys-Elica) [inline Tools/LibFVR/internal/dct.cpp]: Doesn't compile on MSVC (error C2540)

File: Tools/LibFVR/internal/dct.cpp [hunk 2]
```diff
@@ -1,107 +1,598 @@
...
+            for (int i = 0; i <= block_len; i++)
+            {
+                buffer[(block_start + i) * 3] = data[data_offset++];
+            }
         }
 
-        DCT_init = (DCT_cdtor)GetProcAddress(libDct, "??0DCT@@QAE@XZ");
-        DCT_deinit = (DCT_cdtor)GetProcAddress(libDct, "??1DCT@@QAE@XZ");
-        if (!DCT_init || !DCT_deinit)
+        buffer[513 * 3] = 0x7fffffff;
```
Comment (Alys-Elica) [inline Tools/LibFVR/internal/dct.cpp:244]: Index is out of bounds

**Review by Alys-Elica — CHANGES_REQUESTED** [review]
When the above are fixed, the code still crashes (out of bound access)

**Comment by Alys-Elica** [comment]
> @Alys-Elica this isn't quite ready for merge yet, but it works and I wanted to keep you in the loop and get some early feedback.
> 
> I'm much more of a C developer than C++ so if I'm doing anything weird could you perhaps leave a comment there?
> 
> Also, I haven't finished writing up how the DCT works, I'll add some documentation and more comments before merging.
> 
> One thing you'll notice if you compare this implementation with the previous one is that the color saturations are way higher. My suspicion is that the floating to long conversion is different from that of msvcrt, the change happens in the YCbCr -> RGBA conversion.
> 
> Also I'm thinking that perhaps its a good idea to retain the dll based implementation that was already in place and male it conditional, what are your thoughts on that?

Saw your comment a bit late, sorry, I was writing my change request at the same time.

For the C/C++ "weirdiness", the only thing I see for now is the use of malloc. For raw data storage, I like to use std::vector<uint8_t> witch also takes care of the memory management.
There is also already a std::clamp function that exists since C++17.
Also, beware of the unfreed `new`ed objects, they need to be `delete`d when not used anymore. For the objects you use (HuffmanTable and InverseBitStream), I think you could just work with classic non-pointer objects.
If you need infos, you can check the [CPP Reference](cppreference.com) website, it is quite a good doc.

My take on keeping the DLL based DCT is that it could be useful for testing/debugging. For me, the problem using this method is that it locks us in a 32bit Windows only 32bit version. But you are right, keeping the two for now can be useful.

Finally, as I noted previously, I was unable to make it work on my setup, so I didn't fully review it.

**Pull request merged**